source/drivers/MicroBitFileSystem.cpp

```diff
@@ -0,0 +1,1278 @@
+    {
+        bool empty = true;
+        bool deleted = false;
+        uint16_t next;
+
+        for (int i = 0; i < blocksPerPage; i++)
+        {
+            next = getNextFileBlock(page + i);
+            
+            if (next == MBFS_DELETED)
+                deleted = true;
+            
+            else if (next != MBFS_UNUSED)
+            {
+                empty = false;
+                break;
+            }
+        }
+
+        // See if we found one...
+        if (empty)
+        {
+            lastBlockAllocated = page;
+            return getBlock(page);
+        }
+
+        // make note of the first unused but un-erased page we find (if any).
+        if (deleted && !recyclablePage)
+            recyclablePage = page;
+
+        page = (page + blocksPerPage) % fileSystemSize;
+    }
+
+    // No empty pages are available, but we may be able to recycle one.
+    if (recyclablePage)
+    {
+        uint32_t *address = getBlock(recyclablePage);
+        flash.erase_page(address);
+        return address;
+    }
+
+    // Nothing available at all. Use the default.
+    flash.erase_page(defaultScratchPage);
+    return defaultScratchPage;
+}
+
+
+/**
+  * Constructor. Creates an instance of a MicroBitFileSystem.
+  */
+MicroBitFileSystem::MicroBitFileSystem(uint32_t flashStart, int flashPages)
+{
+    // Attempt tp load an existing filesystem, if it exisits
+    init(flashStart, flashPages);
+
+    // If this is the first FileSystem created, so it as the default.
+    if(MicroBitFileSystem::defaultFileSystem == NULL)
+        MicroBitFileSystem::defaultFileSystem = this;
+}
+
+/**
+  * Initialize the flash storage system
+  *
+  * The file system is located dynamically, based on where the program code
+  * and code data finishes. This avoids having to allocate a fixed flash
+  * region for builds even without MicroBitFileSystem.
+  *
+  * This method checks if the file system already exists, and loads it it.
+  * If not, it will determines the optimal size of the file system, if necessary, and format the space 
+  *
+  * @return MICROBIT_OK on success, or an error code.
+  */
+int MicroBitFileSystem::init(uint32_t flashStart, int flashPages)
+{
+    // Protect against accidental re-initialisation
+    if (status & MBFS_STATUS_INITIALISED)
+        return MICROBIT_NOT_SUPPORTED;
+
+    // Validate parameters
+    if (flashPages < 0)
+        return MICROBIT_INVALID_PARAMETER;
+
+    // If we have a zero length, then dynamically determine our geometry.
+    if (flashStart == 0)
+    {
+        // Flash start is on the first page after the programmed ROM contents.
+        // This is: __etext (program code) + static data.
+        // Size of static data is calculated from __data_end__ and __data_start__
+        // (See the linker script)
+
+        flashStart = (uint32_t)&__etext + ((uint32_t)&__data_end__ - (uint32_t)&__data_start__);
+        flashStart = ((uint32_t)flashStart & ~(PAGE_SIZE-1)) + PAGE_SIZE;
```

Comment:
ARMCC has different symbols: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0474i/CHDBIJJD.html

so some conditional compilation required here to support the mbed.org online compiler. Good catch @remay.